core/src/c_api/tiledb.cc

```diff
@@ -302,290 +482,839 @@ int tiledb_basic_array_create(tiledb_ctx_t* ctx, const char* name) {
-      (tiledb_datatype_t*)malloc((attribute_num + 1) * sizeof(int));
-  if (tiledb_array_schema->types_ == nullptr)
-    return TILEDB_OOM;
-  for (int i = 0; i < attribute_num + 1; ++i)
-    tiledb_array_schema->types_[i] = types[i];
+  // Success
+  return TILEDB_OK;
+}
+
+void tiledb_dimension_iter_free(tiledb_dimension_iter_t* dim_it) {
+  if (dim_it == nullptr)
+    return;
+
+  if (dim_it->dim_ != nullptr) {
+    if (dim_it->dim_->dim_ != nullptr)
+      delete (dim_it->dim_->dim_);
+
+    free(dim_it->dim_);
+  }
+
+  free(dim_it);
+}
+
+int tiledb_dimension_iter_done(
+    tiledb_ctx_t* ctx, tiledb_dimension_iter_t* dim_it, int* done) {
+  SANITY_CHECK_CTX(ctx);
+  SANITY_CHECK_DIMENSION_ITER(ctx, dim_it);
+
+  *done = (dim_it->current_dim_ == dim_it->dim_num_) ? 1 : 0;
+
+  return TILEDB_OK;
+}
+
+int tiledb_dimension_iter_next(
+    tiledb_ctx_t* ctx, tiledb_dimension_iter_t* dim_it) {
+  SANITY_CHECK_CTX(ctx);
+  SANITY_CHECK_DIMENSION_ITER(ctx, dim_it);
+
+  ++(dim_it->current_dim_);
+  if (dim_it->dim_ != nullptr) {
+    if (dim_it->dim_->dim_ != nullptr)
+      delete dim_it->dim_->dim_;
 
-  // Set cell val num
-  if (cell_val_num == nullptr) {
-    tiledb_array_schema->cell_val_num_ = nullptr;
-  } else {
-    tiledb_array_schema->cell_val_num_ =
-        (int*)malloc((attribute_num) * sizeof(int));
-    if (tiledb_array_schema->cell_val_num_ == nullptr)
-      return TILEDB_OOM;
-    for (int i = 0; i < attribute_num; ++i) {
-      tiledb_array_schema->cell_val_num_[i] = cell_val_num[i];
-    }
+    if (dim_it->current_dim_ >= 0 && dim_it->current_dim_ < dim_it->dim_num_)
+      dim_it->dim_->dim_ = new tiledb::Dimension(
+          dim_it->array_schema_->array_schema_->dim(dim_it->current_dim_));
+    else
+      dim_it->dim_->dim_ = nullptr;
   }
 
-  // Set cell and tile order
-  tiledb_array_schema->cell_order_ = cell_order;
-  tiledb_array_schema->tile_order_ = tile_order;
+  return TILEDB_OK;
+}
 
-  // Set capacity
-  tiledb_array_schema->capacity_ = capacity;
+int tiledb_dimension_iter_here(
+    tiledb_ctx_t* ctx,
+    tiledb_dimension_iter_t* dim_it,
+    const tiledb_dimension_t** dim) {
+  SANITY_CHECK_CTX(ctx);
+  SANITY_CHECK_DIMENSION_ITER(ctx, dim_it);
 
-  // Set compression
-  if (compression == nullptr) {
-    tiledb_array_schema->compressor_ = nullptr;
-  } else {
-    tiledb_array_schema->compressor_ =
-        (tiledb_compressor_t*)malloc((attribute_num + 1) * sizeof(int));
-    if (tiledb_array_schema->compressor_ == nullptr)
-      return TILEDB_OOM;
-    for (int i = 0; i < attribute_num + 1; ++i)
-      tiledb_array_schema->compressor_[i] = compression[i];
+  *dim = dim_it->dim_;
+
+  return TILEDB_OK;
+}
+
+int tiledb_dimension_iter_first(
+    tiledb_ctx_t* ctx, tiledb_dimension_iter_t* dim_it) {
+  SANITY_CHECK_CTX(ctx);
+  SANITY_CHECK_DIMENSION_ITER(ctx, dim_it);
+
+  dim_it->current_dim_ = 0;
+  if (dim_it->dim_ != nullptr) {
+    if (dim_it->dim_->dim_ != nullptr)
+      delete dim_it->dim_;
```

Comment:
It deletes it as the iterator needs to go back to the first dimension, so a new dimension object must be created (with the constructor that clones the input).